CDAP-4648 Integrated CustomAction class in the WorkflowDriver.

sagarkapare opened the pull request:
JIRA: https://issues.cask.co/browse/CDAP-4648
This PR adds ability to call the `CustomAction` from the `WorkflowDriver`. 
Pending: Removal of usages of `WorkflowAction` from the code base. will be done as a separate PR.

sagarkapare committed:
CDAP-4648 Integrated CustomAction class in the WorkflowDriver.

sagarkapare commented:
Build is running here - http://builds.cask.co/browse/CDAP-DUT4354-1

albertshau commented on cdap-api/src/main/java/co/cask/cdap/api/customaction/CustomActionContext.java:
```diff
@@ -25,7 +25,7 @@
  * Represents runtime context of the {@link CustomAction} in the Workflow.
  */
-public interface CustomActionContext extends RuntimeContext, Transactional, PluginContext, WorkflowInfoProvider {
+public interface CustomActionContext extends RuntimeContext, Transactional, WorkflowInfoProvider {
```
why is PluginContext getting removed? Without it, we cannot use it in apps like Hydrator.

albertshau commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/customaction/BasicCustomActionContext.java:
```diff
@@ -0,0 +1,155 @@
+    try {
+      context.start();
+      runnable.run(datasetCache);
+      context.finish();
```
does this need to be run in a finally? or is it handled by the abortTransaction calls?

sagarkapare replied:
`context.finish()` commits the transaction, which only need to happen if the `run` method succeeds. If there is failure in the execution of the run method, we abort the transaction using `context.abort()`

albertshau commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/CustomActionExecutor.java:
```diff
@@ -75,8 +104,28 @@ private WorkflowAction createAction(BasicWorkflowContext context, InstantiatorFa
 
+  @SuppressWarnings("unchecked")
+  private CustomAction createCustomAction(BasicCustomActionContext context, InstantiatorFactory instantiator,
+                                            ClassLoader classLoader) throws Exception {
```
nit: indentation

sagarkapare replied:
fixed.

albertshau commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/WorkflowProgramInfo.java:
```diff
@@ -61,7 +61,7 @@ public static WorkflowProgramInfo create(Arguments arguments) {
 
-  private WorkflowProgramInfo(String workflowName, String workflowNodeId,
+  WorkflowProgramInfo(String workflowName, String workflowNodeId,
                               String workflowRunId, String programNameInWorkflow, BasicWorkflowToken workflowToken) {
```
nit: indentation

sagarkapare replied:
fixed.

albertshau commented:
is the implementation of PluginContext going to be added back in a separate PR? If so, lgtm after small nitpicks.

sagarkapare commented:
@albertshau Addressed comments. Please take a look again. I will do the implementation of the `PluginContext` in a separate PR where I will make changes to `DataPipelineApp` to include the custom actions.

sagarkapare commented:
Build is green - http://builds.cask.co/browse/CDAP-DUT4354-1

albertshau commented:
lgtm

sagarkapare commented:
Thanks @albertshau for the review. Squashing and merging.
